olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp

```diff
@@ -19,19 +19,107 @@
 
 #include "VolatileLayerClientImpl.h"
 
+#include <olp/core/client/CancellationContext.h>
+#include <olp/dataservice/read/PartitionsRequest.h>
+#include "repositories/DataRepository.h"
+#include "repositories/ExecuteOrSchedule.inl"
+#include "repositories/PartitionsRepository.h"
+
 namespace olp {
 namespace dataservice {
 namespace read {
 
+namespace {
+constexpr auto kLogTag = "VolatileLayerClientImpl";
+} // namespace
+
 VolatileLayerClientImpl::VolatileLayerClientImpl(
     olp::client::HRN catalog, std::string layer_id,
-    olp::client::OlpClientSettings client_settings) {}
+    olp::client::OlpClientSettings client_settings)
+    : catalog_(std::move(catalog)),
+      layer_id_(std::move(layer_id)),
+      client_settings_(std::move(client_settings)),
+      pending_requests_(std::make_shared<PendingRequests>()) {}
 
-VolatileLayerClientImpl::~VolatileLayerClientImpl() = default;
+VolatileLayerClientImpl::~VolatileLayerClientImpl() {
+  pending_requests_->CancelPendingRequests();
+}
 
 olp::client::CancellationToken VolatileLayerClientImpl::GetData(
     DataRequest data_request, Callback callback) {
-  return olp::client::CancellationToken{[]() {}};
+  auto catalog = catalog_;
+  auto client_settings = client_settings_;
+  auto layer_id = layer_id_;
+
+  auto get_data = [=](DataRequest data_request,
+                      olp::client::CancellationContext context,
+                      Callback callback) {
+    if (!data_request.GetDataHandle()) {
+      auto response = repository::PartitionsRepository::GetPartitionById(
+          catalog, layer_id, context, data_request, client_settings);
+
+      if (!response.IsSuccessful()) {
+        if (callback) {
+          callback(response.GetError());
+        }
+        return;
+      }
+
+      const auto& partitions = response.GetResult().GetPartitions();
+      if (partitions.empty()) {
+        OLP_SDK_LOG_INFO_F(kLogTag, "Partition %s not found",
+                           data_request.GetPartitionId()
+                               ? data_request.GetPartitionId().get().c_str()
+                               : "<none>");
+        if (callback) {
+          callback(client::ApiError(client::ErrorCode::NotFound,
+                                    "Partition not found"));
+        }
+        return;
+      }
+
+      data_request.WithDataHandle(partitions.front().GetDataHandle());
+    }
+
+    auto blob_response = repository::DataRepository::GetBlobData(
+        catalog, layer_id, "volatile-blob", data_request, context,
+        client_settings);
+
+    if (callback) {
+      callback(blob_response);
+    }
+  };
+
+  auto pending_requests = pending_requests_;
+  auto add_task = [&](DataRequest& data_request, Callback callback) mutable {
```

Comment:
why mutable?
You only alter pending_requests which is a shared_ptr so you should be able to remove the mutable.